sw/device/lib/base/macros.h

```diff
@@ -162,4 +162,82 @@
  */
 #define OT_UNREACHABLE() __builtin_unreachable()
 
+/**
+ * Defines a local symbol named `kName_` whose address resolves to the
+ * program counter value an inline assembly block at this location would
+ * see.
+ *
+ * The primary intention of this macro is to allow for peripheral tests to be
+ * written that want to assert that a specific program counter reported by the
+ * hardware corresponds to actual code that executed.
+ *
+ * For example, suppose that we're waiting for an interrupt to fire, and want
+ * to verify that it fired in a specific region. We have two volatile globals:
+ * `irq_happened`, which the ISR sets before returning, and `irq_pc`, which
+ * the ISR sets to whatever PC the hardware reports it came from. The
+ * code below tests that the reported PC matches expectations.
+ *
+ * ```
+ * OT_ADDRESSABLE_LABEL(kIrqWaitStart);
+ * enable_interrupts();
+ * while(!irq_happened) {
+ *   wait_for_interrupt();
+ * }
+ * OT_ADDRESSABLE_LABEL(kIrqWaitEnd);
+ *
+ * CHECK(irq_pc >= &kIrqWaitStart && irq_pc <= &IrqWaitEnd);
+ * ```
+ *
+ * Note that this only works if all functions called between the two labels
+ * are actually inlined; if the interrupt fires inside of one of those two
+ * functions, it will appear to not have the right PC, even though it is
+ * logically inside the pair of labels.
+ *
+ * # Volatile Semantics
+ *
+ * This has the same semantics as a `volatile` inline assembly block: it
+ * may be reordered with respect to all operations except other volatile
+ * operations (i.e. volatile assembly and volatile read/write, such as
+ * MMIO). For example, in the following code, `kBefore` and `kAfter` can
+ * wind up having the same address:
+ *
+ * ```
+ * OT_ADDRESSABLE_LABEL(kBefore);
+ * x += 5;
+ * OT_ADDRESSABLE_LABEL(kAfter);
```

Comment:
The semantics of this mechanism are sound when we access a volatile variable between the `kBefore` and `kAfter`. While that is kind of implied in this caveat and in the interrupt example, it is never explicitly stated as a requirement. Maybe consider making that point a bit more explicit?